ext/sockets/sockets.c

```diff
@@ -2755,6 +2774,117 @@ PHP_FUNCTION(socket_addrinfo_explain)
 }
 /* }}} */
 
+#ifdef PHP_WIN32
+
+#include "ext/standard/base64.h"
+
+/* {{{ proto string socket_wsaprotocol_info_export(resource stream, int target_pid)
+   Exports the network socket information suitable to be used in another process. */
+PHP_FUNCTION(socket_wsaprotocol_info_export)
+{
+	WSAPROTOCOL_INFO wi;
+	zval *zsocket;
+	php_socket *socket;
+	zend_long target_pid;
+	zend_string *ret;
+	zend_off_t offset = 0;
+
+	if (zend_parse_parameters(ZEND_NUM_ARGS(), "rl", &zsocket, &target_pid) == FAILURE) {
+		return;
+	}
+	if ((socket = (php_socket *) zend_fetch_resource(Z_RES_P(zsocket), le_socket_name, le_socket)) == NULL) {
+		RETURN_FALSE;
+	}
+
+	if (SOCKET_ERROR == WSADuplicateSocket(socket->bsd_socket, (DWORD)target_pid, &wi)) {
+		DWORD err = WSAGetLastError();
+		LPSTR buf = NULL;
+
+		if (!FormatMessage(
+				FORMAT_MESSAGE_ALLOCATE_BUFFER |
+				FORMAT_MESSAGE_FROM_SYSTEM |
+				FORMAT_MESSAGE_IGNORE_INSERTS,
+				NULL,
+				err,
+				MAKELANGID(LANG_NEUTRAL, SUBLANG_DEFAULT),
+				(LPSTR)&buf,
+			0, NULL)) {
+			php_error_docref(NULL, E_WARNING, "Unable to export WSA protocol info [0x%08lx]", err);
+		} else {
+			php_error_docref(NULL, E_WARNING, "Unable to export WSA protocol info [0x%08lx]: %s", err, buf);
+		}
+
+		RETURN_NULL();
+	}
+
+	/* The simplest way to serialize. The internal data might be of interest,
+		however it is unlikely to ever make sense to have it editable within
+		a script. The struct is suited very well and has same size under both
+		64- and 32-bit. */
+	ret = php_base64_encode((const char *)&wi, sizeof(wi));
```

Comment:
Is there any particular need to serialize it at all? it's just simple binary data

Reply:
I had it in the first variant. Basically can be done that way, but it can be weird especially when the binary data has '\0' bytes. Fe in case some output needs to be checked in tests, etc. No big deal either way otherwise.

Thanks.

Reply:
I'm quite leery about directly exposing (whether encoded or not) such a structure. There's quite a lot of information in there (https://msdn.microsoft.com/en-us/library/windows/desktop/ms741675(v=vs.85).aspx) and it's not immediately obvious to me that nothing bad could happen if the user were to manually modify/construct it before calling the import function.

Reply:
The concern is understandable, of course. The default security mechanism applies, when another process imports the socket. It is for sure possible to craft a `WSAPROTOCOL_INFO` struct. 

The default security also applies, see `DACL`s  https://msdn.microsoft.com/de-de/library/windows/desktop/aa446597(v=vs.85).aspx . Fe it is not possible use a socket created under another account, until it is explicitly set `ACL`s to be allowed. Or it is not possible to duplicate a socket that is not shared (didn't call WSADuplicateSocket). In general, it is hard to do harm to unrelated parts of the system, fe like "stealing" a socket from some another process, on a properly configured system. Same actually when some invalid data gets imported. In this regard, sockets using `SO_REUSEADDR` are a far more insecure solution.

Another approach were to use shared memory. This however introduces another implication by a need to synchronize access. AFAIR Bob was also OK to go by shared memory.

Another approach i could think of were to track the exported structure internally and introduce an extra function that encapsulates the sending process. This introduces a couple of other implications, but is in general doable.

What would you say?

Thanks.